src/main/scala/mesosphere/marathon/core/storage/simple/SimplePersistenceStore.scala

```diff
@@ -0,0 +1,155 @@
+  * is similar to synchronous methods with [[org.apache.curator.framework.api.Backgroundable]] callback however, it
+  * uses a somewhat cleaner [asynchronous DSL](https://curator.apache.org/curator-x-async/index.html).
+  *
+  * Create, update, read and delete operation are offered as akka-stream Flows. Every Flow will offer an output signalising
+  * that the input element was handled successfully. E.g. a [[create]] flow takes a stream of nodes to store and outputs
+  * a stream of node paths indicating that the node was successfully stored.
+  *
+  * For CRUD operations the submitted stream elements are handled with configured parallelism. For create and update
+  * operations the implementation guaranties that if multiple operations update the same path, the order of writes is the
+  * same as in the input stream.
+  *
+  * Transaction support is implemented and four transaction operations are supported: [[CreateOp]], [[UpdateOp]],
+  * [[DeleteOp]] and [[CheckOp]]. All transaction operations has to be successful for the transaction to commit successfully.
+  *
+  * @param factory an instance of [[AsyncCuratorBuilderFactory]]
+  * @param parallelism parallelism level for CRUD operations
+  * @param ec execution context
+  */
+class SimplePersistenceStore(factory: AsyncCuratorBuilderFactory, parallelism: Int = 10)(implicit ec: ExecutionContext)
+  extends PersistenceStore with StrictLogging {
+
+  // format: OFF
+  private[this] val createMetric          = Metrics.counter(ServiceMetric, getClass, "create")
+  private[this] val readMetric            = Metrics.counter(ServiceMetric, getClass, "read")
+  private[this] val updatedMetric         = Metrics.counter(ServiceMetric, getClass, "update")
+  private[this] val deleteMetric          = Metrics.counter(ServiceMetric, getClass, "delete")
+  private[this] val childrenMetric        = Metrics.counter(ServiceMetric, getClass, "children")
+  private[this] val existsMetric          = Metrics.counter(ServiceMetric, getClass, "exists")
+  private[this] val transactionMetric     = Metrics.counter(ServiceMetric, getClass, "transaction")
+  private[this] val transactionCountMetric = Metrics.counter(ServiceMetric, getClass, "transactionCount")
+
+
+  // Helper stages for logging and metrics
+  def logNode(prefix: String): Flow[Node, Node, NotUsed]                = Flow[Node].map{ n => logger.debug(s"$prefix${n.path}"); n }
+  def logPath(prefix: String): Flow[String, String, NotUsed]            = Flow[String].map{ p => logger.debug(s"$prefix${p}"); p }
+  def metric[T](counter: Counter, times: Long = 1): Flow[T, T, NotUsed] = Flow[T].map{ e => counter.increment(times); e}
+  // format: ON
+
+  override def create: Flow[Node, String, NotUsed] =
+    Flow[Node]
+      .via(logNode("Creating a node at "))
+      .via(metric(createMetric))
+      .groupBy(parallelism, node => Math.abs(node.path.hashCode) % parallelism)
+      .map(node => factory
```

Comment:
Hm, if you group by a number what is map receiving? Why do we do the grouping anyways? What is the return type?